### ISSUE TITLE
Array.prototype.copyWithin should throw when trying to delete a non-configurable property

### ISSUE THREAD

**Issue by sethbrenith** [Microsoft/ChakraCore#4356]
I wrote the following test case (which could go in ES6ArrayApi.js), and `copyWithin` fails to throw where expected.

```javascript
    {
        name: "Array.copyWithin() should respect overridden `length` property on TypedArray instance",
        body: function () {
            var x = new Int32Array([1,2,3]);
            Object.defineProperty(x, "length", { value: 4 });
            assert.throws(function() { Array.prototype.copyWithin.call(x, 1, 2); }, TypeError,
                "We should get a TypeError when trying to delete an item of TypedArray");
            assert.areEqual("1,3,3,", Array.prototype.join.call(x, ","),
                "We should have already modified previous items before throwing");

            x[1] = 2;
            Object.setPrototypeOf(x, { "3": 5 });
            assert.throws(function() { Array.prototype.copyWithin.call(x, 1, 2); }, TypeError,
                "We should still get a TypeError because [[HasProperty]] on an Integer Indexed Exotic Object returns false for out-of-bounds values");
            assert.areEqual("1,3,3,", Array.prototype.join.call(x, ","),
                "We should not read prototype properties of TypedArray by index");

            x[2] = 4;
            Int32Array.prototype.copyWithin.call(x, 0, 1);
            assert.areEqual("3,4,4,", Array.prototype.join.call(x, ","),
                "TypedArray copyWithin ignores overridden length property and uses internal [[ArrayLength]]");
        }
    },
```

It seems that the issue stems from the fact that several places in `JavascriptArray::CopyWithinHelper` call `OP_DeleteElementI` or `DeleteItem` but don't wrap the call in a `ThrowTypeErrorOnFailure`.

This bug is possible because correct usage of `PropertyOperation_ThrowOnDeleteIfNotConfig` is super confusing: it looks like it should make `DeleteItem` throw, but it doesn't. Either we should remove that enum value or we should update `DeleteItem` and friends respect it.

**Comment by digitalinfinity** [comment]
@irinayat-MS is this the same issue as #4781?

**Comment by irinayat-MS** [comment]
@digitalinfinity at the moment I don't think so and this one should be easier to fix, but if I find later that the cause is the same and can be fixed in one place, I'll dupe it to #4781. The main difference is that delete shouldn't throw in non-strict mode while copyWithin (splice, shift, etc.) should.
